reservation_units/migrations/0105_refactor_pricing_state.py

```diff
@@ -0,0 +1,49 @@
+# Generated by Django 5.0.8 on 2024-09-02 12:44
+
+from django.db import migrations, models
+from django.db.models import Count
+
+
+def remove_duplicate_date_pricings(apps, schema_editor):
+    apps.get_model("reservation_units", "ReservationUnit")
+    ReservationUnitPricing = apps.get_model("reservation_units", "ReservationUnitPricing")
+
+    duplicates_qs = (
+        ReservationUnitPricing.objects.values("reservation_unit_id", "begins")
+        .annotate(count=Count("begins"))
+        .filter(count__gt=1)
+    )
+    for duplicate in duplicates_qs:
+        # Delete all but the latest pricing with the same reservation_unit_id and begins, as it will be the active one
+        for pricing in ReservationUnitPricing.objects.filter(
+            reservation_unit_id=duplicate["reservation_unit_id"],
+            begins=duplicate["begins"],
+        ).order_by("-pk")[1:]:
+            pricing.delete()
```

Comment:
I think this will work, but there is a more succinct way to express this using window functions:
```python
ReservationUnitPricing.objects.annotate(
    row=Window(
        RowNumber(),
        partition_by=("reservation_unit", "begins"),
        order_by="-pk",
    ),
).filter(row__gt=1).delete()
```

This will annotate a row number for each pricing so that row number starts from 1 for each unique reservation unit and begins date pair, starting from the last created. Therefore, we can just bulk remove anything that has a row number greater than 1, since they would be duplicate.

Should test this still if you're going to use it.

Reply:
Essentially the same thing, but this looks a lot cleaner :D

Reply:
Too bad that that this doesn't work as-is due to database level restrictions:
```
django.db.utils.ProgrammingError: window functions are not allowed in WHERE
LINE 1: DELETE FROM "reservation_unit_pricing" WHERE ROW_NUMBER() OV...
```
I decided to keep the original duplicate removal since it works fine and in the end doesn't really matter that much 😅